### ISSUE TITLE
Large memory increase when performing operation on diagonal of kronecker product vs kronecker product directly

### ISSUE THREAD

**Issue by Nebulised** [jax-ml/jax#29705]
### Description

When getting the gradient of functions involving the diagonal of the kronecker product there is a significant increase in vram usage. In some rough tests anywhere from 2x -10x .

In the example below in my case changing the operation from being performed on the matrix vs the diagonal of the matrix changes my VRAM usage from ~160MB to 2200MB.

I also printed out the jaxpr instructions and noticed the number of lines of instructions for the non diagonal case was 102. For the diagonal case (high memory usage case)  was 241.

The VRAM usage only seems to increase when taking the gradient and when using the kronecker. If I just do the mean of the diagonal/the mean on a parameter array passed in to the loss function, no change in memory usage. 

```python
import os
os.environ["XLA_PYTHON_CLIENT_PREALLOCATE"] = "0"
import jax.numpy as jnp
import jax

import optax


HIGH_MEMORY = False # CHANGE THIS TO ENABLE/DISABLE HIGH MEMORY USAGE


def compute_loss(params):
    loss = jnp.kron(params, jnp.eye(200))
    return jnp.sum(jnp.diagonal(loss)) if HIGH_MEMORY else jnp.sum(loss)


key = jax.random.PRNGKey(42)

start_learning_rate = 1e-1
optimizer = optax.adam(start_learning_rate)
params = jax.random.normal(key=key, shape=(100,100))

opt_state = optimizer.init(params)

@jax.jit
def step(params, opt_state):
    loss, grads = jax.value_and_grad(compute_loss)(params)
    updates, opt_state = optimizer.update(grads, opt_state)
    params = optax.apply_updates(params, updates)
    return params, opt_state, loss



for _ in range(256):
    step(params, opt_state)
```

### System info (python version, jaxlib version, accelerator, etc.)

jax:    0.6.0
jaxlib: 0.6.0
numpy:  2.2.4
optax : 0.2.4
python: 3.12.3 (main, Jun 18 2025, 17:59:45) [GCC 13.3.0]
device info: NVIDIA GeForce GTX 1080 Ti-2, 2 local devices"
process_count: 1
Ubuntu (24.04)
NVIDIA-SMI 550.144.03             Driver Version: 550.144.03     CUDA Version: 12.4

**Comment by jakevdp** [comment]
The main difference here is that `sum` is a single XLA primitive, while extracting the diagonal of an array is a compound operation built from a number of XLA primitives. So it's not surprising that using `sum(diag(x))` rather than `sum(x)` would result in a longer jaxpr, more operations, and more memory consumption.

That said, 2x to 10x does seem like a lot, and it may point to our implementaiton of `jnp.diagonal` being sub-optimial. Currently the [default implementation](https://github.com/jax-ml/jax/blob/78e57d01f582eab5a901e892341f208f444cc414/jax/_src/numpy/lax_numpy.py#L7271-L7281) is based on a gather operation, with a [reduce-based operation](https://github.com/jax-ml/jax/blob/78e57d01f582eab5a901e892341f208f444cc414/jax/_src/numpy/lax_numpy.py#L7291-L7299) via mosaic on GPU. Since you report running on GPU, it seems like this latter implementation is the relevant one.

I think this was contributed by @voznesenskym, so perhaps he can comment here?

**Comment by Nebulised** [comment]
Cheers for the prompt reply. Yes, I found it strange to be such a large increase. I have just run it on a different machine on another GPU and get similar results. 

I have also just run the example with CPU only enabled and it doesn't appear to suffer the same problem although it is harder for me to tell by looking at the RAM usage. 

As a workaround for this specific case multiplying by the identity to zero out non diagonal matrices does solve the memory issue. 

I have also noticed when performing an operation leading to a larger array inside the loss function, in some cases some kind of optimisation is missed? When using jax.random.normal without the kronecker product,  large memory usage occurs regardless of jnp.diagonal usage. However, maybe this is due to difference in operations as it is obviously not a like for like comparison. I have included this script below in case it's of any use. 

```python

import os
os.environ["XLA_PYTHON_CLIENT_PREALLOCATE"] = "0"
import jax.numpy as jnp
import jax
import optax


HIGH_MEMORY = False # CHANGE THIS TO ENABLE/DISABLE HIGH MEMORY USAGE
USE_KRONECKER = False # CHANGE THIS TO ENABLE/DISABLE USING KRONECKER VS EXP OF IDENTITY



key = jax.random.PRNGKey(42)
def compute_loss(params):
    if USE_KRONECKER:
        loss = params * jnp.kron(jax.random.normal(key=jax.random.key(0), shape=(100, 100)), jnp.eye(200,200))
    else:
        loss = params * jnp.exp(jnp.eye(20000))
        # loss = params * jax.random.normal(key=jax.random.key(0), shape=(20000,20000)) # THIS CAUSES HIGH MEMORY USAGE REGARDLESS
    return jnp.sum(jnp.diagonal(loss)) if HIGH_MEMORY else jnp.sum(loss)




start_learning_rate = 1e-1
optimizer = optax.adam(start_learning_rate)
params = jax.random.normal(key=key, shape=(1,))

opt_state = optimizer.init(params)

@jax.jit
def step(params, opt_state):
    loss, grads = jax.value_and_grad(compute_loss)(params)
    updates, opt_state = optimizer.update(grads, opt_state)
    params = optax.apply_updates(params, updates)
    return params, opt_state, loss



for _ in range(256):
    params, opt_state, loss = step(params, opt_state)
```

**Comment by RRiva** [comment]
Hi, I have very limited experience with JAX, but your code reminds me of [this discussion](https://discourse.julialang.org/t/how-do-i-implement-the-fastest-tensor-contractions/111007/6) on the Julia Discourse. I suggest skipping the many multiplications with 0 and 1 and just write a for loop. The memory reduction should be obvious and using `@jit` or `jax.lax.fori_loop` should take care of the speed.
